src/NHibernate/Impl/SqlQueryImpl.cs

```diff
@@ -320,5 +327,37 @@ protected internal override IEnumerable<ITranslator> GetTranslators(ISessionImpl
 			var sqlQuery = this as ISQLQuery;
 			yield return new SqlTranslator(sqlQuery, sessionImplementor.Factory);
 		}
+
+		public ISQLQuery AddSynchronizedQuerySpace(string querySpace)
+		{
+			addedQuerySpaces.Value.Add(querySpace);
+			return this;
+		}
+
+		public ISQLQuery AddSynchronizedEntityName(string entityName)
+		{
+			var persister = session.Factory.GetEntityPersister(entityName);
+			foreach (var querySpace in persister.QuerySpaces)
+			{
+				addedQuerySpaces.Value.Add(querySpace);
+			}
+			return this;
+		}
+
+		public ISQLQuery AddSynchronizedEntityClass(System.Type entityType)
+		{
+			var persister = session.Factory.GetEntityPersister(entityType.FullName);
+			foreach (var querySpace in persister.QuerySpaces)
+			{
+				addedQuerySpaces.Value.Add(querySpace);
+			}
+			return this;
+		}
+
+		public IReadOnlyCollection<string> GetSynchronizedQuerySpaces()
+		{
+			return addedQuerySpaces.IsValueCreated ? addedQuerySpaces.Value : new HashSet<string>();
```

Comment:
The `new HashSet<string>()` kills the purpose of the `Lazy<>` here.

Reply:
I think the `Lazy<>` is not worth it here. For having it useful, `GetSynchronizedQuerySpaces()` would have to yield `null` instead when it is not initialized. Would it be a private method, why not, but as an interface method, better yield an empty collection.